File: Medium/3. Longest Substring Without Repeating Characters/src/main.java
public class main {

    public static int lengthOfLongestSubstring(String s) {
        if(s.length() == 0)
            return 0;

        int i_low = 0;
        int i_high = 0;

        int max_len = 1;
        int cur_len = 1;

        while(i_high < s.length() - 1) {
            i_high ++;
            char c_high = s.charAt(i_high);

            boolean is_added = true;
            for(int i = i_low; i < i_high; i++) {
                if(c_high == s.charAt(i)) {
                    //start new segment
                    max_len = Math.max(max_len, i_high - i_low);

                    i_low ++;
                    i_high = i_low + 1;

                    is_added = false;
                    cur_len = 1;
                    break;
                }
            }
            if(is_added) {
                cur_len += 1;
            }
        }

        max_len = Math.max(max_len, cur_len);

        return max_len;
    }

    public static void main(String[] args) {
//        assert(lengthOfLongestSubstring("abcabcbb") == 3);
//        assert(lengthOfLongestSubstring("bbbbb") == 1);
//        assert(lengthOfLongestSubstring("pwwkew") == 3);
//        assert(lengthOfLongestSubstring("") == 0);
//        assert(lengthOfLongestSubstring("au") == 2);
        assert(lengthOfLongestSubstring("dvdf") == 3);
    }
}

File: Medium/3. Longest Substring Without Repeating Characters/python/main.py

# abcabcbb
#i   
# j       
def lengthOfLongestSubstring(s: str) -> int:
    l, charSet, longest_ss = 0, set(), 0 # Contains all unique characters, also checking for a character in the set is O(1)
    # We don't know how to initialize the right pointer so we have for loop and while loop
    for r in range(len(s)):
        # While the s[r] in our charset, we need to shrink the left window and remove the left character from the charSet and shrink left window
        while s[r] in charSet:
            charSet.remove(s[l])
            l += 1
        charSet.add(s[r]) # After removed all duplicates, the s[r] doesn't appear in charset so we can safely add it
        longest_ss = max(longest_ss, r - l + 1)
    return longest_ss
            
if __name__ == "__main__":
    assert lengthOfLongestSubstring("abcabcbb") == 3
    assert lengthOfLongestSubstring("bbbbb") == 1
    assert lengthOfLongestSubstring("pwwkew") == 3
    assert lengthOfLongestSubstring("") == 0
